### ISSUE TITLE
[BUG] Bloated MoE expert checkpoint files with ZeRO 2

### ISSUE THREAD

**Issue by clumsy** [microsoft/DeepSpeed#2389]
**Describe the bug**
It appears that with ZeRO 2 MoE layer expert parameters are replicated across all ranks by sharing a single tensor storage (with different offsets).
Since all MoE layer experts are stored in individual files for all ranks the default `torch.save()` behavior stores the underlying tensor's storage (even though a given expert is only covering a fraction of the storage).

Thus we end up with the following file structure per layer N:
```
layer_N_expert_E_mp_rank_00_model_states.pt
...
layer_N_expert_2_mp_rank_00_model_states.pt
layer_N_expert_1_mp_rank_00_model_states.pt
layer_N_expert_0_mp_rank_00_model_states.pt
```

Where _each_ file has the storage for all E experts across layer N, but the tensors in the files are only using a fraction of it.
When `torch.load()`ed each end up being tensors with independent storage.

It seems like using `.clone().detach()` when saving MoE expert tensors to individual files should address the issue.

But I was wondering about the reason behind making each expert and individual file for all ranks. E.g. if we stored a single file for expert layer for all ranks then the storage would have been reused for the loaded tensors and this single file would not be bloated.

I'll provide a test shortly, willing to contribute the fix.

**To Reproduce**

**Expected behavior**
Checkpoint files should be compact, no redundancy.

**ds_report output**
Please run `ds_report` to give us details about your setup.

**Screenshots**

**System info (please complete the following information):**
 - OS: Amazon Linux 4.14.252-195.483.amzn2.x86_64
 - GPU count and types: 8x A100 on a single machine
 - Interconnects (if applicable): single machine
 - Python version: 3.8
 - Any other relevant info about your setup: DeepSpeed 0.6.0, 0.7.4

**Launcher context**
DeepSpeed runner/launcher

**Docker context**
Custom docker image

**Additional context**
Using MoE layers, ZeRO 2

**Comment by clumsy** [comment]
This issue only happens when there are multiple MoE layers and half-precision ZeRO 2 is used where all distributed parallel's group parameters' storages are flattened using [this logic](https://github.com/microsoft/DeepSpeed/blame/e419f7cbcda6339f712e1657cd55ed0800ec6171/deepspeed/runtime/zero/stage_1_and_2.py#L551).

Now a single expert's model parameters are at an offset in this shared storage, hence when `torch.save()` stores it dumps it all.

**Comment by tjruwase** [comment]
@clumsy, if you are still able to contribute a fix that will be greatly appreciated. Thanks!

**Comment by clumsy** [comment]
> @clumsy, if you are still able to contribute a fix that will be greatly appreciated. Thanks!

@tjruwase Yes, I have the fix and test changes ready. Just need to finish running all tests and I'll cut a PR.
